File: mobile/lib/models/asistencia.dart
/// Espejo de `app/schemas/asistencia.py` para el lado del Aprendiz
/// (`GET /asistencias/mias`).
///
/// El móvil es de solo lectura por arquitectura (ver ARQUITECTURA_MOBILE.md),
/// y en asistencia eso coincide con la regla de negocio: la asistencia la
/// certifica el instructor que dictó la clase, así que acá no hay nada que
/// escribir aunque el cliente pudiera.
library;

enum EstadoAsistencia { presente, tardanza, excusa, ausente }

EstadoAsistencia _estadoDesde(String valor) => switch (valor) {
      'presente' => EstadoAsistencia.presente,
      'tardanza' => EstadoAsistencia.tardanza,
      'excusa' => EstadoAsistencia.excusa,
      _ => EstadoAsistencia.ausente,
    };

extension EstadoAsistenciaX on EstadoAsistencia {
  String get etiqueta => switch (this) {
        EstadoAsistencia.presente => 'Presente',
        EstadoAsistencia.tardanza => 'Tarde',
        EstadoAsistencia.excusa => 'Excusa',
        EstadoAsistencia.ausente => 'Ausente',
      };
}

class SesionAsistida {
  final int idAsistencia;
  final DateTime fechaSesion;
  final EstadoAsistencia estado;
  final String? referenciaExcusa;
  final String? resultadoDescripcion;
  final String? instructorNombre;
  final String? ambienteNombre;
  final String horaInicio;
  final String horaFin;

  const SesionAsistida({
    required this.idAsistencia,
    required this.fechaSesion,
    required this.estado,
    required this.horaInicio,
    required this.horaFin,
    this.referenciaExcusa,
    this.resultadoDescripcion,
    this.instructorNombre,
    this.ambienteNombre,
  });

  factory SesionAsistida.fromJson(Map<String, dynamic> json) => SesionAsistida(
        idAsistencia: json['idAsistencia'] as int,
        fechaSesion: DateTime.parse(json['fechaSesion'] as String),
        estado: _estadoDesde(json['estado'] as String),
        referenciaExcusa: json['referenciaExcusa'] as String?,
        resultadoDescripcion: json['resultadoDescripcion'] as String?,
        instructorNombre: json['instructorNombre'] as String?,
        ambienteNombre: json['ambienteNombre'] as String?,
        horaInicio: (json['horaInicio'] as String?) ?? '',
        horaFin: (json['horaFin'] as String?) ?? '',
      );
}

class ResumenAsistencia {
  final int registradas;
  final int presente;
  final int tardanza;
  final int excusa;
  final int ausente;

  /// Sobre sesiones REGISTRADAS, no sobre las programadas del trimestre: el
  /// sistema solo sabe de las clases a las que ya les pasaron lista.
  final double porcentaje;

  const ResumenAsistencia({
    required this.registradas,
    required this.presente,
    required this.tardanza,
    required this.excusa,
    required this.ausente,
    required this.porcentaje,
  });

  factory ResumenAsistencia.fromJson(Map<String, dynamic> json) => ResumenAsistencia(
        registradas: json['registradas'] as int? ?? 0,
        presente: json['presente'] as int? ?? 0,
        tardanza: json['tardanza'] as int? ?? 0,
        excusa: json['excusa'] as int? ?? 0,
        ausente: json['ausente'] as int? ?? 0,
        porcentaje: (json['porcentaje'] as num?)?.toDouble() ?? 100.0,
      );
}

class MiAsistencia {
  final ResumenAsistencia resumen;
  final List<SesionAsistida> sesiones;

  const MiAsistencia({required this.resumen, required this.sesiones});

  factory MiAsistencia.fromJson(Map<String, dynamic> json) => MiAsistencia(
        resumen: ResumenAsistencia.fromJson(json['resumen'] as Map<String, dynamic>),
        sesiones: (json['sesiones'] as List<dynamic>? ?? [])
            .map((s) => SesionAsistida.fromJson(s as Map<String, dynamic>))
            .toList(),
      );
}

File: backend/app/schemas/asistencia.py
from datetime import date, datetime, time
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict

# Los cuatro que acepta la columna `estado`. Literal y no Enum de Postgres,
# mismo criterio que el resto del proyecto (ver Notificacion.tipo).
EstadoAsistencia = Literal["presente", "tardanza", "excusa", "ausente"]


class AprendizDeSesion(BaseModel):
    """Una fila de la lista que el instructor marca. `estado` viene en null
    cuando todavía no se le pasó lista a esa persona ese día — el cliente lo
    muestra como "sin marcar", que no es lo mismo que "ausente"."""

    idUsuario: UUID
    nombre: str
    numeroDocumento: str | None = None
    rolEnFicha: str | None = None

    estado: EstadoAsistencia | None = None
    horaMarcacion: datetime | None = None
    referenciaExcusa: str | None = None


class SesionAsistencia(BaseModel):
    """Todo lo que la pantalla de pasar lista necesita de una sesión."""

    idHorario: int
    fechaSesion: date
    fichaCodigo: str | None = None
    resultadoDescripcion: str | None = None
    ambienteNombre: str | None = None
    horaInicio: time
    horaFin: time

    aprendices: list[AprendizDeSesion]

    # Cuándo y quién pasó lista por última vez. null = sesión sin registrar.
    registradaEn: datetime | None = None
    registradaPor: str | None = None


class MarcaAsistencia(BaseModel):
    idUsuarioAprendiz: UUID
    estado: EstadoAsistencia
    referenciaExcusa: str | None = None


class RegistroAsistencia(BaseModel):
    """La lista completa de una sesión, no marca por marca: pasar lista es
    un acto único y así el servidor puede reemplazar lo que hubiera."""

    idHorario: int
    fechaSesion: date
    marcas: list[MarcaAsistencia]


class AsistenciaDeAprendiz(BaseModel):
    """Una fila del historial del aprendiz."""

    model_config = ConfigDict(from_attributes=True)

    idAsistencia: int
    idHorario: int
    fechaSesion: date
    estado: EstadoAsistencia
    referenciaExcusa: str | None = None

    resultadoDescripcion: str | None = None
    instructorNombre: str | None = None
    ambienteNombre: str | None = None
    horaInicio: time
    horaFin: time


class ResumenAsistencia(BaseModel):
    """Las cifras del encabezado. `porcentaje` es sobre sesiones
    REGISTRADAS, no sobre las programadas del trimestre: el sistema solo
    sabe de las clases a las que alguien le pasó lista."""

    registradas: int
    presente: int
    tardanza: int
    excusa: int
    ausente: int
    porcentaje: float


class MiAsistencia(BaseModel):
    resumen: ResumenAsistencia
    sesiones: list[AsistenciaDeAprendiz]
